lib/say_cases.rb

```diff
@@ -0,0 +1,59 @@
+class SayCase < OpenStruct
+  def test_name
+    'test_%s' % description.downcase.tr(' ', '_')
+  end
+
+  def workload
+    [
+      "question = '#{input}'",
+      indent(4, assertion),
+    ].join("\n")
+  end
+
+  def skipped
+    index.zero? ? '# skip' : 'skip'
+  end
+
+  private
+
+  def indent(size, lines)
+    lines.lines.each_with_object('') { |line, obj| obj << ' ' * size + line }
+  end
+
+  def assertion
+    return error_assertion unless expected
+    return message_assertion if message
+
+    "assert_equal(#{expected}, SayProblem.new(question).answer)"
+  end
+
+  def error_assertion
+    [
+      'assert_raises ArgumentError do',
+      indent(2, 'SayProblem.new(question).answer'),
+      'end',
+    ].join("\n")
+  end
+
+  def message_assertion
+    [
+      'answer = SayProblem.new(question).answer',
+      "message = \"#{message % '#{answer}'}\"",
+      "assert_equal(#{expected}, answer, message)",
+    ].join("\n")
+  end
+end
+
+class SayCase::PreProcessor
+  class << self
+    def call(row)
+      row
+    end
+  end
+end
+
+SayCases = proc do |data|
+  JSON.parse(data)['cases'].map.with_index do |row, i|
+    SayCase.new(SayCase::PreProcessor.call(row).merge(index: i))
```

Comment:
Because this exercise doesn't require preprocessing, the `PreProcessor` just returns the provided `row`.

You can remove the `PreProcessor` class and merge the index directly into `row`